### DISCUSSION TITLE
Phase Field - Grain growth - ACGrGrPoly

### DISCUSSION THREAD

**Discussion by EtienneSwerim** [idaholab/moose#16550] [Q&A Modules: Phase field] [answered]
Hi!

Could somebody help me with the following?

I am trying to set up the simplest possible grain growth model (three grains) in the Phase Field module by combining parts of existing example files.

I append my input file (myTricrystalTripleJunction.docx); it runs well to the end but the results do not seem physically correct: the three order parameters even out to 0.3 over the whole model instead of being either 1 (grain) or 0 (no grain).

Can it be so that I need to modify the ACCGrGrPoly.C file? I noticed that the residual it returns is only d_F/d_eta. Firstly, there is no test function term and secondly there is no parameter L like in the manual (Basic Phase Field Equations). L only appear later in the file for the offDiagonal jacobian terms. I append it in the Appendix.docx file.

Very grateful for help!

Best Regards
Etienne

[Appendix.docx](https://github.com/idaholab/moose/files/5721000/Appendix.docx)

[myTricrystalTripleJunction.docx](https://github.com/idaholab/moose/files/5720992/myTricrystalTripleJunction.docx)

**Comment by laagesen** [comment, marked as answer]
Hi Etienne,

`ACGrGrPoly` is a class derived from `ACGrGrBase`, which is in turn derived from `ACBulk`. Multiplying by the Allen-Cahn mobility `L` actually is done in `ACBulk` in the function `precomputeQpResidual`. `ACBulk` is a then a derived class of `KernelValue`. Multiplying by the test function is done in `KernelValue`.

To get started with a grain-growth simulation, I'd suggest using one of the example input files in the MOOSE repository and changing parameters for your problem, rather than trying to create your own input file. The simplest to start with is probably `moose/modules/phase_field/examples/grain_growth/grain_growth_2D_random.i`. You can change the number of order parameters using the global parameter `op_num` and change initial conditions to what you prefer. This example uses the Action system to add all the required kernels and materials to obtain the evolution equations.

As far as the non-physical behavior you were seeing, I looked quickly at your input file and I think the problem is at least partially coming from the fact that you are specifying `v = 'eta0 eta1 eta2'` as a GlobalParam. Each ACGrGrPoly kernel needs `v` specified, and it needs to be a list of all the other order parameters in the system besides the nonlinear variable for that kernel. For example in the ACGrGrPoly kernel for `variable = eta0`, you would specify `v = 'eta1 eta2'`. With `v` specified as a GlobalParam you end up with extra terms in your evolution equations which do not give the correct physics. For these reasons I'd suggest you start with the example from the paragraph above which uses the action system to fill in the correct kernels and parameters specified to each one.

**Reply by EtienneSwerim** [reply]
Dear Larry,

Thank you very much for your detailed answer! I really appreciate you took the time to go through the relation between all the classes you mentioned.

I have now looked at grain_growth_2D_random.i and even at grain_growth_2D_voronoi.i and I will now try to understand how the Action system works.

I have also modified my input file according to your advice. I define a different v-vector for each ACGrGrPoly and it works perfectly! Again, I am very grateful for your help.

Best Regards
Etienne

**Reply by laagesen** [reply]
Great glad that helped!

**Comment by EtienneSwerim** [comment]
Hi,

Could I get a little help with the appended example file: grain_growth_2D_random.i?

As I understand from the Grain Growth Model section of the Moose Manual, order parameters and kernels are automatically created by the Action system called GrainGrowthAction. I even found the files GrainGrowthAction.h and GrainGrowthAction.C.

What I do not understand is:

1) Where, in the grain_growth_2D_random.i, are the GrainGrowthAction files called? The name GrainGrowthAction is not present in the file. I guess these files are called implicitly but how?

2) When it comes to variables, only the name PolycrystalVariables is defined but neither a type nor an action, see below. I noticed that Moose crashes if the name PolycrystalVariables is changed and I noticed that there is a folder called PolycrystalVariables 
(moose/module/phase_field/doc/content/syntax/Variables/PolycrystalVariables)

This folder contains a single file called index.md which contain the following single line: 

!template load file=stubs/moose_system.md.template name=PolycrystalVariables syntax=/Variables/PolycrystalVariables

a) Is this index.md file called by the variable name PolycrystalVariables in the grain_growth_2D_random.i?

b) What does this index.md file do? I noticed there were a lot of theses files in the moose folder for example one below PolycrystalKernel and one below PolycrystalRandomIC and I do not understand what these two do either.

Very grateful for help!

Best Regards
Etienne

[Variables]
  # Variable block, where all variables in the simulation are declared
  [./PolycrystalVariables]
    # Custom action that created all of the grain variables and sets their initial condition
  [../]
[]

[grain_growth_2D_random.txt](https://github.com/idaholab/moose/files/5966053/grain_growth_2D_random.txt)

**Reply by laagesen** [reply]
Hi Etienne,

1) The kernels are created by the `PolycrystalKernel` action, which is called in the `Kernels` block of that file. The variables are created by the `PolycrystalVariables` action. This particular example doesn't use the `GrainGrowthAction`. The source code files for these actions are in the same directory as the `GrainGrowthAction` files.

2) The files with the .md extension are markdown files for the MOOSE documentation system. They are used to create the web-based documentation you see at, for example, https://mooseframework.inl.gov/modules/phase_field/index.html. You shouldn't need to look at these unless you are creating or modifying documentation.

**Reply by EtienneSwerim** [reply]
Hi Larry,
Thanks a lot for your answer! I understand now that these action files are called by the name of the variables and the name of the kernel.
Best regards
Etienne
